src/Symfony/Component/DependencyInjection/Secrets/JweEntry.php

```diff
@@ -0,0 +1,131 @@
+    protected function compact()
+    {
+        $valuesForCompaction = array(
+            $this->base64url_encode($this->header),
+            sodium_bin2base64($this->encryptedKey, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->iv, SODIUM_BASE64_VARIANT_URLSAFE),
+            sodium_bin2base64($this->cipherText, SODIUM_BASE64_VARIANT_URLSAFE),
+            $this->base64url_encode($this->authTag)
+        );
+
+        return implode(".", $valuesForCompaction);
+    }
+
+    protected function decryptedSecret(string $keyPair)
+    {
+        $plaintextCek = sodium_crypto_box_seal_open(
+            $this->encryptedKey,
+            $keyPair
+        );
+
+        if (sodium_crypto_aead_aes256gcm_is_available()) {
+            $decrypted = sodium_crypto_aead_aes256gcm_decrypt(
+                $this->cipherText,
+                $this->authTag,
+                $this->iv,
+                $plaintextCek
+            );
+            if ($decrypted === false) {
+                throw new RuntimeException(sprintf(
+                    "Unable to decrypt secrets. Verify the configured key pair"
+                ));
+            }
+        } else {
+            //TODO add alt decryption method
+            $decrypted = "";
+        }
+
+        return $decrypted;
+    }
+
+    protected function hydrate(string $compactedEntry)
+    {
+        //TODO add validations
+        list($header, $encryptedKey, $iv, $cipherText, $authTag) = explode(".", $compactedEntry);
+        $this->header = $this->base64url_decode($header);
+        $this->encryptedKey = sodium_base642bin($encryptedKey, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->iv = sodium_base642bin($iv, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->cipherText = sodium_base642bin($cipherText, SODIUM_BASE64_VARIANT_URLSAFE);
+        $this->authTag = $this->base64url_decode($authTag);
+    }
+
+
+    private function setEncryptedValues(string $secret, string $pubKey)
+    {
+        $this->header = $this->generateHeader();
+        $rawCek = sodium_crypto_aead_aes256gcm_keygen();
+        $encryptedCek = sodium_crypto_box_seal(
+            $rawCek,
+            $pubKey
+        );
+
+        $this->encryptedKey = $encryptedCek;
+        $nonce = random_bytes(SODIUM_CRYPTO_AEAD_AES256GCM_NPUBBYTES);
+        $this->iv = $nonce;
+        $additionalData = $this->generateHeader();
+        if (sodium_crypto_aead_aes256gcm_is_available()) {
+            $ciphertext = sodium_crypto_aead_aes256gcm_encrypt(
+                $secret,
+                $additionalData,
+                $nonce,
+                $rawCek
+            );
+        } else {
+            $ciphertext = "";
+            //TODO: add alternative encryption method
+        }
+
+        //TODO: figure out if this conversion is needed
+        //mb_convert_encoding($additionalData, "ASCII");
+        $this->authTag = $additionalData;
+        $this->cipherText = $ciphertext;
+    }
```

Comment:
This class will be a tough one :). Where did you get all this logic from? Obviously, because it's security, this needs to be done perfectly. And so, ideally, we're building all of this logic based on something "tried-and-true".

Reply:
this is all based on translating https://tools.ietf.org/html/rfc7516 into php and reading a bunch of these docs- https://paragonie.com/book/pecl-libsodium. The feedback I got from security folks was to use to try and use aes256gcm if possible. I was having a hard time finding reputable/current JWE libs that were worth bringing in another dependency for, and since libsodium is built in to php 7.2 and SecurityExtension already uses it i thought it was a decent candidate. If there is an industry standard lib for this though I completely agree it should be used